src/Fantomas.CoreGlobalTool/Program.fs

```diff
@@ -96,24 +98,38 @@ let rec allFiles isRec path =
 /// Format a source string using given config and write to a text writer
 let processSourceString isFsiFile s (tw : Choice<TextWriter, string>) config =
     let fileName = if isFsiFile then "/tmp.fsi" else "/tmp.fsx"
-    async {
-        let! formatted = CodeFormatter.FormatDocumentAsync(fileName, SourceOrigin.SourceString s, config,
-                                                           FakeHelpers.createParsingOptionsFromFile fileName,
-                                                           FakeHelpers.sharedChecker.Value)
+    let writeResult (formatted: string) =
         match tw with
         | Choice1Of2 tw -> tw.Write(formatted)
         | Choice2Of2 path -> File.WriteAllText(path, formatted)
+
+    async {
+        let! formatted = s |> FakeHelpers.formatContentAsync config fileName
+
+        match formatted with
+        | FakeHelpers.FormatResult.Formatted(_, formattedContent) ->
+            formattedContent |> writeResult
+        | FakeHelpers.FormatResult.Unchanged(_) ->
+            s |> writeResult
+        | FakeHelpers.FormatResult.Error(_, ex) ->
+            raise <| ex
     }
     |> Async.RunSynchronously
 
 /// Format inFile and write to text writer
-let processSourceFile inFile (tw : TextWriter) config = 
-    let s = File.ReadAllText(inFile)
+let processSourceFile inFile (tw : TextWriter) config =
     async {
-        let! formatted = CodeFormatter.FormatDocumentAsync(inFile, SourceOrigin.SourceString s, config,
-                                                           FakeHelpers.createParsingOptionsFromFile inFile,
-                                                           FakeHelpers.sharedChecker.Value)
-        tw.Write(formatted)
+        let! formatted = inFile |> FakeHelpers.formatFileAsync config
+
+        match formatted with
+        | FakeHelpers.FormatResult.Formatted(_, formattedContent) ->
+            tw.Write(formattedContent)
+        | FakeHelpers.FormatResult.Unchanged(_) ->
+            inFile
```

Comment:
I know this is probably the current behaviour but does it make sense to write out the same file?

Reply:
You are right. I think the reason it was done like this originally is because your `output file` could be different than your `input file` and they still wanted the unchanged files to be copied over.

I am okay with restricting this a bit, if you prefer, by identifying if the underlying TextWriter destination is the same as the source, to avoid unnecessary writes. 

Or I could give a pass to the CLI logic altogether and enumerate a bunch of use cases that it currently supports, but don't make much sense.

Reply:
Ah yes, indeed if the input and output path is different it should write the file even if no formatting was necessary.
In theory, you could also check if the formatted file differs from the output path and don't write if it isn't the case.

So if you have input path A and output path B, compare formatted A with B (if it exists) before writing to path B.

> Or I could give a pass to the CLI logic altogether and enumerate a bunch of use cases that it currently supports, but don't make much sense.

I would keep it simple: if no formatting was applied no file should be written.